### PR TITLE
[tools/shoestring] fix: add support for Ubuntu 20.04

### PR DESCRIPTION
**Pull request by Wayonb**
problem: Ubuntu 20.04 ships with an older version of Python and Openssl which caused failures when running shoestring.
               1. Openssl 1.1.x output is different from Openssl 3.x
               2. Python 3.8, ``shutil.move`` does not support a path object.
solution: 1. Normalize the Openssl output so it works for both version
               2. Python 3.8 convert the path objects to string.

### CODE REVIEW

File: tools/shoestring/shoestring/healthagents/rest_https_certificate.py [hunk 1]
```diff
@@ -24,25 +26,45 @@ def _openssl_run_sclient_verify(hostname, test_args):
...
 	] + test_args, command_input='Q', show_output=False)
 
+	# verify the certificate is valid
+	for line in lines:
+		line = line.strip()
+		if line.startswith('verify error:'):
+			# normalize error message
+			if '1.1.1' in openssl_executor.version():
+				line = line.replace('self signed', 'self-signed')
+			return False, line
```
Comment (Jaguar0625) [inline tools/shoestring/shoestring/healthagents/rest_https_certificate.py:36]: is it possible to add test for this case?

File: tools/shoestring/shoestring/healthagents/rest_https_certificate.py [hunk 2]
```diff
@@ -24,25 +26,45 @@ def _openssl_run_sclient_verify(hostname, test_args):
...
+
+	with tempfile.TemporaryDirectory() as temp_directory:
+		temp_file = Path(temp_directory) / 'cert.txt'
+		with open(temp_file, 'w', encoding='utf-8') as cert_file:
```
Comment (Jaguar0625) [inline tools/shoestring/shoestring/healthagents/rest_https_certificate.py:40]: cert_outfile ?

Comment (Jaguar0625) [reply tools/shoestring/shoestring/healthagents/rest_https_certificate.py:40]: wt

File: tools/shoestring/shoestring/healthagents/rest_https_certificate.py [hunk 3]
```diff
@@ -24,25 +26,45 @@ def _openssl_run_sclient_verify(hostname, test_args):
...
+			cert_file.writelines(lines)
+
+		lines = openssl_executor.dispatch([
+			'x509',
+			'-inform',
+			'pem',
```
Comment (Jaguar0625) [inline tools/shoestring/shoestring/healthagents/rest_https_certificate.py:46]: consider putting these on same line since pem is input to inform

File: tools/shoestring/shoestring/healthagents/rest_https_certificate.py [hunk 4]
```diff
@@ -24,25 +26,45 @@ def _openssl_run_sclient_verify(hostname, test_args):
...
+		], command_input='Q', show_output=False)
+
+	searching_for_date = [r'Not Before: (.*)', r'Not After : (.*)']
 	collect_dates = False
 	collected_dates = []
 	for line in lines:
 		line = line.strip()
-		if collect_dates:
-			res = re.search(r'NotBefore: (.*); NotAfter: (.*)', line)
+		if collect_dates and searching_for_date:
```
Comment (Jaguar0625) [inline tools/shoestring/shoestring/healthagents/rest_https_certificate.py:58]: searching_for_date => date_patterns?

File: tools/shoestring/shoestring/healthagents/rest_https_certificate.py [hunk 5]
```diff
@@ -24,25 +26,45 @@ def _openssl_run_sclient_verify(hostname, test_args):
...
 			collect_dates = True
 
-		if line.startswith('Verify return code:'):
-			if 'Verify return code: 0 (ok)' == line:
-				return True, collected_dates
-
-			return False, line
-
-	return False, 'could not parse s_client response'
+	return (True, collected_dates) if 2 == len(collected_dates) else (False, 'could not parse s_client response')
```
Comment (Jaguar0625) [inline tools/shoestring/shoestring/healthagents/rest_https_certificate.py:67]: original version would collect multiple certificate date ranges

but idk if we need that anymore, so new version is probably fine

cc: @gimre-xymcity

File: tools/shoestring/shoestring/internal/OpensslExecutor.py [hunk 6]
```diff
@@ -37,6 +37,6 @@ def dispatch(self, args, command_input=None, show_output=True):
 
 			if 0 != process.returncode:
 				formatted_command_line = ' '.join(command_line)
-				raise RuntimeError(f'{formatted_command_line} exited with {process.returncode}')
+				raise RuntimeError(f'{formatted_command_line} exited with {process.returncode}, stdout: {stdout_lines}')
```
Comment (Jaguar0625) [inline tools/shoestring/shoestring/internal/OpensslExecutor.py:40]: stdout:\n ?

File: tools/shoestring/tests/commands/test_renew_certificates.py [hunk 7]
```diff
@@ -82,7 +82,7 @@ async def test_can_renew_node_certificate():
 
 
 async def test_can_renew_node_certificate_with_ca_password():
-	await _assert_can_renew_node_certificate('abc')
+	await _assert_can_renew_node_certificate('abcd')
```
Comment (Jaguar0625) [inline tools/shoestring/tests/commands/test_renew_certificates.py:85]: why this change? is password limit different?

File: tools/shoestring/shoestring/healthagents/rest_https_certificate.py (hunk 1 above)
Comment (Wayonb) [reply tools/shoestring/shoestring/healthagents/rest_https_certificate.py:36]: Add a testcase for the normalize message scenario?
There is a test for self-signed -  ``test_validate_fails_when_certificate_is_self_signed``

File: tools/shoestring/tests/commands/test_renew_certificates.py (hunk 7 above)
Comment (Wayonb) [reply tools/shoestring/tests/commands/test_renew_certificates.py:85]: yea, the password limit is different between versions.

File: tools/shoestring/shoestring/healthagents/rest_https_certificate.py (hunk 1 above)
Comment (Jaguar0625) [reply tools/shoestring/shoestring/healthagents/rest_https_certificate.py:36]: i mean error branch

File: tools/shoestring/shoestring/internal/OpensslExecutor.py (hunk 6 above)
Comment (Jaguar0625) [reply tools/shoestring/shoestring/internal/OpensslExecutor.py]: can remove space after \n now :)

File: tools/shoestring/tests/healthagents/test_rest_https_certificate.py [hunk 8]
```diff
@@ -53,15 +55,23 @@ async def _dispatch_validate(test_args=None):
 	HealthAgentContext = namedtuple('HealthAgentContext', ['hostname', 'test_args'])
 	context = HealthAgentContext('localhost', test_args or [])
 
-	await asyncio.to_thread(_validate_thread, context)
+	await asyncio.get_running_loop().run_in_executor(None, _validate_thread, context)
 
 
 async def test_validate_fails_when_certificate_is_self_signed(server, caplog):  # pylint: disable=redefined-outer-name, unused-argument
+	# Arrange:
+	error_message = 'HTTPS certificate looks invalid: verify error:num=18:self-signed certificate'
```
Comment (Jaguar0625) [inline tools/shoestring/tests/healthagents/test_rest_https_certificate.py:63]: expected_error_message?

**Review by Jaguar0625 — APPROVED** [review]

File: tools/shoestring/shoestring/healthagents/rest_https_certificate.py [hunk 9]
```diff
@@ -24,25 +26,40 @@ def _openssl_run_sclient_verify(hostname, test_args):
...
+		with open(temp_file, 'wt', encoding='utf-8') as cert_outfile:
+			cert_outfile.writelines(lines)
+
+		lines = openssl_executor.dispatch([
+			'x509',
+			'-inform', 'pem',
+			'-noout',
+			'-text',
+			'-in', temp_file
+		], command_input='Q', show_output=False)
```
Comment (gimre-xymcity) [inline tools/shoestring/shoestring/healthagents/rest_https_certificate.py:46]: does `x509` command require `command_input`? (it should be fire and forget, that's why I doubt that)

File: tools/shoestring/shoestring/healthagents/rest_https_certificate.py (hunk 5 above)
Comment (gimre-xymcity) [reply tools/shoestring/shoestring/healthagents/rest_https_certificate.py:62]: I think we do.
IIRC concatenated pem file has CA followed by node cert, and IIRC `x509` only parses FIRST certificate :| (that's why I've used `s_client` output)
There was some other command that could be used, I just don't remember what it was.
